### PR TITLE
Add LSP server support

### PR DESCRIPTION
**Pull request by craigmc08**
Adds support for using [waspls](https://github.com/wasp-lang/wasp/tree/main/waspls) in the extension. The extension now looks in PATH for `waspls` (or uses the user configured path, if set in the extension settings) for the executable and runs it. The code that deals with this is in `src/extension.ts`.

This does not bundle `waspls`, so separate installation of the language server is still needed.

Also, it updates the textmate grammar to match the latest wasp language and adds code snippets for pages, queries, actions, and entities.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -6,12 +6,20 @@
...
+            "runtimeExecutable": "${execPath}",
             "args": [
-                "--extensionDevelopmentPath=${workspaceFolder}"
-            ]
+                "--extensionDevelopmentPath=${workspaceRoot}"
+            ],
+            "outFiles": [
+                "${workspaceRoot}/out/**/*.js"
+            ],
+            "preLaunchTask": {
```
Comment (Martinsos) [inline .vscode/launch.json:19]: Why is this needed?

File: README.md [hunk 2]
```diff
@@ -31,6 +46,13 @@ VSCode needs .json, not .yaml -> use `npm run build` to generate .json from .yam
    which is at the end, applying correct scopes.
 7. Repeat step 4.
 
+For features implemented in typescript, the entry point for the project is `src/extension.ts`.
+Use `npm run compile-ts` to compile the typescript code, or `npm run watch` to 
+automatically recompile on changes.
+
+Use `npm run compile` to compile typescript code and convert the TextMate grammar
```
Comment (Martinsos) [inline README.md:53]: Aha, so `npm run compile` does everything, while `compile-ts` and `compile-yaml` do specific stuff. Ok makes sense!

File: package.json [hunk 3]
```diff
@@ -43,14 +54,53 @@
...
+        "wasp.server.executable": {
+          "scope": "resource",
+          "type": "string",
+          "description": "Path to wasp language server executable. Tries to use waspls on PATH if not set."
+        },
+        "wasp.server.useOutputPanelForLogging": {
+          "scope": "resource",
+          "type": "boolean",
+          "default": true,
+          "description": "Send logs from the wasp language server to the output panel. If enabled, the log file setting will be ignored"
```
Comment (Martinsos) [inline package.json:71]:
```suggestion
          "description": "Send logs from the wasp language server to the output panel. If enabled, the log file setting will be ignored."
```

File: snippets/wasp.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+  "Page": {
+    "prefix": "page",
+    "description": "Create a new page and matching",
```
Comment (Martinsos) [inline snippets/wasp.json:4]: missing smth?

File: src/extension.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import { execFile } from 'child_process';
+
+let client: LanguageClient;
+
+const outputChannel = window.createOutputChannel('Wasp Language Server');
+
+export async function activate(context: ExtensionContext) {
+  console.log('..Extension "vscode-wasp" is now active..');
+
+  const config = workspace.getConfiguration('wasp');
```
Comment (Martinsos) [inline src/extension.ts:20]: Where is this string `'wasp'` coming from?

File: src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+  // Compute path to `waspls` executable, based on settings
+  const userDefinedExecutablePath = config.server.executable;
+  let executablePath = userDefinedExecutablePath === '' ? 'waspls' : userDefinedExecutablePath;
```
Comment (Martinsos) [inline src/extension.ts:24]: Is this safer than
```
let executablePath = userDefinedExecutablePath || 'waspls'
```
Maybe this makes it more robust?

File: src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    return;
+  }
+
+  let folders = workspace.workspaceFolders;
+  if (folders) {
+    let folder = folders[0];
+    if (folder) {
+      executablePath = executablePath.replace('${workspaceFolder}', folder.uri.path).replace('${workspaceRoot}', folder.uri.path);
+    }
+  }
```
Comment (Martinsos) [inline src/extension.ts:43]: Why is this needed? Who is defining these vars?

File: src/extension.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    } else {
+      if (userDefinedExecutablePath === '') {
+        window.showErrorMessage('No `waspls` executable is available in the VSCode PATH.');
+        return;
+      } else {
+        window.showInformationMessage(`The user defined executable path couldn't be run: [${executablePath}].`);
+      }
+    }
+  }
+
```
Comment (Martinsos) [inline src/extension.ts:62]: A lot of code for just figuring out the executable path and checking if it is valid!
Maybe we could separate it into a separate function.

File: src/extension.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+// Runs the executable in `version` mode and checks that it:
+// - Exits successfuly
+// - Exits within 1 second
+//
+// If it fails either of these checks, a status is returned to show that the
+// executable path is not valid.
+async function checkExecutable(executablePath: string): Promise<Status> {
+  const execPromise = promisify(execFile)
+    (executablePath, ['version'], { timeout: 2000, windowsHide: true })
```
Comment (Martinsos) [inline src/extension.ts:133]: Why 2s here?

**Review by Martinsos — APPROVED** [review]
LGTM @craigmc08 ! Left some minor comments but nothing big, approved, so feel free to merge when ready!

File: .vscode/launch.json (hunk 1 above)
Comment (craigmc08) [reply .vscode/launch.json:19]: This will watch for changes and recompile the typescript when you test the extension from in VSCode (using F5).

File: snippets/wasp.json (hunk 4 above)
Comment (craigmc08) [reply snippets/wasp.json:4]: Fixed.

File: src/extension.ts (hunk 5 above)
Comment (craigmc08) [reply src/extension.ts:20]: From `package.json`, in `contributes.configuration`. It's based on the property names. I agree that it's kind of weird, but this is the way vscode makes you do it.

File: src/extension.ts (hunk 6 above)
Comment (craigmc08) [reply src/extension.ts:24]: It's probably the same in this case. I've switched it to your suggestion, since that's usually more common in JS code I think.

File: src/extension.ts (hunk 7 above)
Comment (craigmc08) [reply src/extension.ts:43]: VSCode has some built-in variables like these that they substitute in in their configuration properties, but they don't provide any API for resolving the variables. Extensions commonly interpolate in these variables.

File: src/extension.ts (hunk 9 above)
Comment (craigmc08) [reply src/extension.ts:133]: It's an arbitrary timeout. I picked it because 2s seemed like a reasonable amount of time to wait before deciding the executable isn't working.

File: src/extension.ts (hunk 8 above)
Comment (craigmc08) [reply src/extension.ts:62]: Refactored into a separate function.

**Comment by craigmc08** [comment]
@Martinsos Ready for merge! I'm not a contributor on the repo so I'm not able to do it.

File: src/extension.ts (hunk 5 above)
Comment (Martinsos) [reply src/extension.ts:20]: Maybe mention it in the comment.

File: src/extension.ts (hunk 6 above)
Comment (Martinsos) [reply src/extension.ts]: Good! I think it is useful because it also covers any stuff like undefined, null and similar. Sometimes that is good, sometimes not, but here I think it is good.

File: src/extension.ts (hunk 7 above)
Comment (Martinsos) [reply src/extension.ts]: Ok! Maybe leave a small comment.

**Comment by Martinsos** [comment]
@craigmc08 left couple more tiny comments and gave you needed permissions to merge!
